// graphql-transformer-core/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
require("./polyfills/Object.assign");
var TransformerContext_1 = require("./TransformerContext");
exports.TransformerContext = TransformerContext_1.default;
var Transformer_1 = require("./Transformer");
exports.Transformer = Transformer_1.default;
var GraphQLTransform_1 = require("./GraphQLTransform");
var collectDirectives_1 = require("./collectDirectives");
exports.collectDirectiveNames = collectDirectives_1.collectDirectiveNames;
var stripDirectives_1 = require("./stripDirectives");
exports.stripDirectives = stripDirectives_1.stripDirectives;
var amplifyUtils_1 = require("./util/amplifyUtils");
exports.buildAPIProject = amplifyUtils_1.buildProject;
exports.uploadAPIProject = amplifyUtils_1.uploadDeployment;
exports.readProjectSchema = amplifyUtils_1.readSchema;
exports.migrateAPIProject = amplifyUtils_1.migrateAPIProject;
exports.revertAPIMigration = amplifyUtils_1.revertAPIMigration;
exports.readProjectConfiguration = amplifyUtils_1.readProjectConfiguration;
__export(require("./errors"));
exports.default = GraphQLTransform_1.default;
